### PR TITLE
[WIP] Fix ellipsoid cartesianToCartographic

### PR DESCRIPTION
**Pull request by xintongxia**

### CODE REVIEW

File: modules/math/src/geospatial/ellipsoid.js [hunk 1]
```diff
@@ -283,7 +288,7 @@ export default class Ellipsoid {
    */
   geodeticSurfaceNormal(cartesian, result = [0, 0, 0]) {
     vec3.multiply(result, cartesian, this._oneOverRadiiSquared);
-    return vec3.normalize(result, result);
+    return vec3.normalize(new Vector3(), result);
```
Comment (ibgreen) [inline modules/math/src/geospatial/ellipsoid.js]: Why this change? you want the result to be a `Vector3()`?

But this function is supposed to return result if provided, not a new `Vector3`

Change the default argument to `result=new Vector3()` instead of `result=[0,0,0]`?

File: modules/math/src/geospatial/ellipsoid.js [hunk 2]
```diff
@@ -325,19 +330,19 @@ export default class Ellipsoid {
...
 
-    const normal = cartesianToCartographicN;
-    this.geodeticSurfaceNormal(p, normal);
+    const normal = this.geodeticSurfaceNormal(point, cartesianToCartographicN);
 
     const h = cartesianToCartographicH;
-    h.copy(cartesian).subtract(p);
+    h.copy(cartesian).subtract(point);
 
-    const longitude = Math.atan2(n.y, n.x);
```
Comment (ibgreen) [inline modules/math/src/geospatial/ellipsoid.js]: If `n` is an instance of `Vector3()`, `.x` and `.y` should work...

File: modules/math/src/geospatial/ellipsoid.js [hunk 3]
```diff
@@ -325,19 +330,19 @@ export default class Ellipsoid {
...
-    const latitude = Math.asin(n.z);
+    const longitude = Math.atan2(normal[1], normal[0]);
+    const latitude = Math.asin(normal[2]);
     const height = Math.sign(vec3.dot(h, cartesian)) * vec3.length(h);
 
-    result[0] = longitude;
-    result[1] = latitude;
+    result[0] = longitude * RAD_TO_DEGREE;
```
Comment (ibgreen) [inline modules/math/src/geospatial/ellipsoid.js]: OR `import {degrees as toDegrees} from 'math.gl';`

File: modules/math/src/geospatial/helpers/scale-to-geodetic-surface.js [hunk 4]
```diff
@@ -22,15 +22,15 @@ const scaleToGeodeticSurfaceGradient = new Vector3();
...
-  const positionY = cartesian.y;
-  const positionZ = cartesian.z;
+  const positionX = cartesian.x || 0;
+  const positionY = cartesian.y || 0;
+  const positionZ = cartesian.z || 0;
 
-  const oneOverRadiiX = oneOverRadii.x;
-  const oneOverRadiiY = oneOverRadii.y;
-  const oneOverRadiiZ = oneOverRadii.z;
+  const oneOverRadiiX = oneOverRadii.x || 0;
```
Comment (ibgreen) [inline modules/math/src/geospatial/helpers/scale-to-geodetic-surface.js]: Do we really need these defaults? zero values here seem like an incorrect ellipse, maybe we should `assert` instead?

File: modules/math/src/geospatial/helpers/scale-to-geodetic-surface.js [hunk 5]
```diff
@@ -40,29 +40,33 @@ export default function scaleToGeodeticSurface(cartesian, ellipsoid, result) {
   const squaredNorm = x2 + y2 + z2;
   const ratio = Math.sqrt(1.0 / squaredNorm);
 
+  if (!Number.isFinite(ratio)) {
```
Comment (ibgreen) [inline modules/math/src/geospatial/helpers/scale-to-geodetic-surface.js]: When does this happen? Why did we have to add this? Add a comment?

File: modules/math/src/geospatial/helpers/scale-to-geodetic-surface.js [hunk 6]
```diff
@@ -40,29 +40,33 @@ export default function scaleToGeodeticSurface(cartesian, ellipsoid, result) {
...
+
   // As an initial approximation, assume that the radial intersection is the projection point.
-  const intersection = scaleToGeodeticSurfaceIntersection;
+  const intersection = new Vector3();
   intersection.copy(cartesian).scale(ratio);
 
   // If the position is near the center, the iteration will not converge.
   if (squaredNorm < centerToleranceSquared) {
-    return !isFinite(ratio) ? undefined : vec3.copy(result, intersection);
+    return intersection;
```
Comment (ibgreen) [inline modules/math/src/geospatial/helpers/scale-to-geodetic-surface.js]: We need to populate and return `result` if that is what the user supplied.

File: modules/math/src/geospatial/helpers/scale-to-geodetic-surface.js [hunk 7]
```diff
@@ -40,29 +40,33 @@ export default function scaleToGeodeticSurface(cartesian, ellipsoid, result) {
   const squaredNorm = x2 + y2 + z2;
   const ratio = Math.sqrt(1.0 / squaredNorm);
 
+  if (!Number.isFinite(ratio)) {
+    return undefined;
+  }
+
   // As an initial approximation, assume that the radial intersection is the projection point.
-  const intersection = scaleToGeodeticSurfaceIntersection;
+  const intersection = new Vector3();
```
Comment (ibgreen) [inline modules/math/src/geospatial/helpers/scale-to-geodetic-surface.js]: We could pre-allocate this `Vector3` to avoid having to allocate an object here.

File: modules/math/src/geospatial/helpers/scale-to-geodetic-surface.js [hunk 8]
```diff
@@ -40,29 +40,33 @@ export default function scaleToGeodeticSurface(cartesian, ellipsoid, result) {
...
-  const oneOverRadiiSquaredY = oneOverRadiiSquared.y;
-  const oneOverRadiiSquaredZ = oneOverRadiiSquared.z;
+  const oneOverRadiiSquaredX = oneOverRadiiSquared.x || 0;
+  const oneOverRadiiSquaredY = oneOverRadiiSquared.y || 0;
+  const oneOverRadiiSquaredZ = oneOverRadiiSquared.z || 0;
 
   // Use the gradient at the intersection point in place of the true unit normal.
   // The difference in magnitude will be absorbed in the multiplier.
-  const gradient = scaleToGeodeticSurfaceGradient;
+  const gradient = new Vector3();
```
Comment (ibgreen) [inline modules/math/src/geospatial/helpers/scale-to-geodetic-surface.js]: Preallocate `tempGradient`...

File: modules/math/test/geospatial/ellipsoid.spec.js [hunk 9]
```diff
@@ -95,7 +95,6 @@ test('Ellipsoid#geodeticSurfaceNormal works with a result parameter', t => {
   const ellipsoid = Ellipsoid.WGS84;
   const result = new Vector3();
   const returnedResult = ellipsoid.geodeticSurfaceNormal(spaceCartesian, result);
-  t.ok(returnedResult === result);
```
Comment (ibgreen) [inline modules/math/test/geospatial/ellipsoid.spec.js]: This test is there to verify that you are actually returning the result object if supplied.

File: modules/math/test/geospatial/ellipsoid.spec.js [hunk 10]
```diff
@@ -125,35 +123,39 @@ test('Ellipsoid#cartesianToCartographic works without a result parameter', t =>
...
 test('Ellipsoid#cartesianToCartographic works close to center', t => {
-  const expected = new Vector3(9.999999999999999e-11, 1.0067394967422763e-20, -6378137.0);
+  const expected = new Vector3(
+    (9.999999999999999e-11 * 180) / Math.PI,
+    (1.0067394967422763e-20 * 180) / Math.PI,
+    -6378137.0
+  );
   const returnedResult = Ellipsoid.WGS84.cartesianToCartographic(new Vector3(1e-50, 1e-60, 1e-70));
-  t.equals(returnedResult, expected);
+  tapeEqualsEpsilon(t, returnedResult, expected, MathUtils.EPSILON8);
```
Comment (ibgreen) [inline modules/math/test/geospatial/ellipsoid.spec.js]: Did we add this epsilon, or was it present in the original tests?

File: modules/math/src/geospatial/helpers/scale-to-geodetic-surface.js (hunk 5 above)
Comment (xintongxia) [reply modules/math/src/geospatial/helpers/scale-to-geodetic-surface.js]: When very close to center, `[0, 0, 0]`, the value `ratio = Math.sqrt(1.0 / squaredNorm);` is infinite.
